### PR TITLE
feat: add GraphQL group and groups query

### LINKED ISSUE #3914: Create groups GraphQL query
**Summary**
Create a `groups` query in GraphQL.

**Implementation**
The `groups` query should have the following signature
```
groups(shopId: ID!): GroupConnection
```
It has one argument, `shopId` which is required, and returns a `GroupConnection` associated with that shopId if the user requesting the groups has permissions to access those groups.

**Resources**
[Github UserConnection docs](https://developer.github.com/v4/object/userconnection/)

## Acceptance Criteria
Using GraphQL, I can use the `groups(shopId: ID!)` query to get a `GroupConnection` by shopId.

### LINKED ISSUE #3915: Create group GraphQL query
**Summary**
Create a `group` query in GraphQL.

**Implementation**
The `group` query should have the following signature
```
groups(id: ID!): Group
```
It has one argument, `id` which is required, and returns the `Group` associated with that id if the user requesting the `Group` has permissions to access that `Group` and the `Group` exists.

## Acceptance Criteria
Using GraphQL, I can use the `group(id: ID!)` query to get a `Group` by id.

### PR DESCRIPTION
**Pull request by mikemurray**
Resolves #3914
Resolves #3915 
Impact: **minor**  
Type: **feature**

## Issue

Add GraphQL group and groups query to be able to query for a single group by ID, or multiple groups by shop ID.

## Solution

Docs are WIP with more details here: https://github.com/reactioncommerce/reaction-docs/pull/564

tl;dr

1. Build the method that fetches data, placed in the account package for groups
1. Build the resolvers for the group and groups queries in Accounts
1. Build resolvers for createdAt and shop
1. Update shop and group schemas for a group connection

## Breaking changes

n/a

## Testing

1. Start the app and sign in
1. open `localhost:3000/graphiql`
1. Run the following queries to get results

```

{
  group(id: "<base64IdOfAGroup>") {
    _id
    name
    slug
  }
}
```

```
  shop(id: "cmVhY3Rpb24vc2hvcDpKOEJocTN1VHRkZ3daeDNyeg==") {
  	_id,
    groups(first: 2) {
      totalCount
      pageInfo {
        endCursor
        startCursor
        hasNextPage
        hasPreviousPage
      }
      edges {
        cursor
        node {	
          name
          shop {
            _id
          }
        }
      }
    }
	}
}
```

```
  groups(shopId: "cmVhY3Rpb24vc2hvcDpKOEJocTN1VHRkZ3daeDNyeg==", first: 2) {
    totalCount
    pageInfo {
      endCursor
      startCursor
      hasNextPage
      hasPreviousPage
    }
    edges {
    	cursor
      node {	
        name
        shop {
          _id
        }
      }
    }
  }
}
```

### CODE REVIEW

File: imports/test-utils/helpers/getFakeMongoCursor.js [hunk 1]
```diff
@@ -0,0 +1,33 @@
+/**
+ * @name getFakeMongoCursor
+ * @param {String} collectionName - name of the collection
+ * @param {Any} results - results to be returned as part of the cursor
+ * @return {Object} fake cursor
+ */
+export default function getFakeMongoCursor(collectionName, results) {
```
Comment (aldeed) [inline imports/test-utils/helpers/getFakeMongoCursor.js]: The same function is in `imports/plugins/core/graphql/server/resolvers/shop/Shop/administrators.test.js` and `imports/plugins/core/graphql/server/resolvers/account/Query/administrators.test.js`. Delete it from those files and import from here instead.

It would also be nice if this could replace the `mockCursor` I have in some test files. The problem is that those typically need the functions to be jest mocks rather than just plain functions. So for example the tests need:

```
limit: jest.fn().mockName("cursor.limit").mockReturnValue(cursor)
```

whereas for devserver we just need `() => cursor` because jest isn't loaded.

I think you could make this function a bit smarter to handle both use cases.

```
export default function getFakeMongoCursor(collectionName, results) {
  const cursor = {
    ns: `meteor.${collectionName}`,
    // ... etc. for all non-function props
  };
  const isTesting = typeof jest !== "undefined";

  if (isTesting) {
    cursor.limit = jest.fn().mockName("cursor.limit").mockReturnValue(cursor);
  } else {
    cursor.limit = () => cursor;
  }

  // and so on, for all props that are functions

  return cursor;
}
```

These are the places that have mock cursors for testing that could use this instead:

- imports/plugins/core/graphql/server/resolvers/util/applyPaginationToMongoCursor.test.js
- imports/plugins/core/graphql/server/resolvers/util/getCollectionFromCursor.test.js
- imports/plugins/core/graphql/server/resolvers/util/getPaginatedResponse.test.js

Do you mind giving that a try?

File: imports/plugins/core/graphql/server/resolvers/account/Query/groups.js [hunk 2]
```diff
@@ -0,0 +1,24 @@
...
+ * @name groups
+ * @method
+ * @summary find and return the administrators (users with "admin" or "owner" role) for a shop
+ * @param {Object} _ - unused
+ * @param {Object} args - an object of all arguments that were sent by the client
+ * @param {String} args.shopId - id of user to query
+ * @param {String} args.after - Connection argument
+ * @param {String} args.before - Connection argument
+ * @param {Number} args.first - Connection argument
+ * @param {Number} args.last - Connection argument
```
[the remark this answers is not in the document]
Comment (mikemurray) [reply imports/plugins/core/graphql/server/resolvers/account/Query/groups.js]: Maybe we could use an `@see` to link to a general place that has the additional properties.

File: imports/test-utils/helpers/getFakeMongoCursor.js (hunk 1 above)
Comment (mikemurray) [reply imports/test-utils/helpers/getFakeMongoCursor.js]: looking into it

File: imports/plugins/core/graphql/server/resolvers/account/Query/groups.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (aldeed) [reply imports/plugins/core/graphql/server/resolvers/account/Query/groups.js]: If you can make that work

[the remark this answers is not in the document]
Comment (mikemurray) [reply imports/plugins/core/graphql/server/resolvers/account/Query/groups.js]: Did a quick test, this is what I have for it using jsdoc `@typedef`

```
/**
 * Arguments passed by the client for a query
 * @typedef {Object} graphql-args - an object of all arguments that were sent by the client
 * @property {String} args.after - Connection argument
 * @property {String} args.before - Connection argument
 * @property {Number} args.first - Connection argument
 * @property {Number} args.last - Connection argument
 */
```

```
/**
 * @name administrators
 * @method
 * @summary find and return the administrators (users with "admin" or "owner" role) for a shop
 * @param {Object} shop - The shop object returned by the parent resolver
 * @param {...graphql-args} args - an object of all arguments that were sent by the client {@link graphql-args}
 * @param {Object} context - an object containing the per-request state
 * @return {Object[]} Promise that resolves with array of user account objects
 */
```

reaction_docs_-_api__global

reaction_docs_-_api__global

File: imports/test-utils/helpers/getFakeMongoCursor.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+      }
+    }
+  };
+
+  const isTesting = typeof jest !== "undefined";
+
+  if (isTesting) {
+    cursor.clone = jest.fn().mockName("cursor.clone").mockReturnValue({
+      count: jest.fn().mockName("cursor.clone.count").mockReturnValueOnce(results.length)
+    });
```
Comment (aldeed) [inline imports/test-utils/helpers/getFakeMongoCursor.js]: Change this to

```js
cursor.clone = jest.fn().mockName("cursor.clone").mockImplementation(() => cursor);
```

And then add:

```js
cursor.count = jest.fn().mockName("cursor.count").mockReturnValue(results.length)
```

And then you don't have to do `mockCursor.clone = () => ({ count: mockCount });` in `applyPaginationToMongoCursor.test.js`

(Make the same change in the non-jest code. Have clone return `cursor` and add `cursor.count` func.)

File: imports/plugins/core/graphql/server/resolvers/account/Query/groups.js (hunk 2 above)
[the remark this answers is not in the document]
Comment (mikemurray) [reply imports/plugins/core/graphql/server/resolvers/account/Query/groups.js]: The jsdoc typedef has to live in a file somewhere. I currently have it in `schemas/index.js`. Do you think there's a better place to place it?

```
/**
 * Arguments passed by the client for a query
 * @typedef {Object} graphql-args - an object of all arguments that were sent by the client
 * @property {String} args.after - Connection argument
 * @property {String} args.before - Connection argument
 * @property {Number} args.first - Connection argument
 * @property {Number} args.last - Connection argument
 * @property {Number} args.sortBy - Connection argument. Check schema for allowed values.
 * @property {Number} args.sortOrder - Connection argument
 */

```

[the remark this answers is not in the document]
Comment (aldeed) [reply imports/plugins/core/graphql/server/resolvers/account/Query/groups.js]: Maybe in resolvers/index.js instead since the resolvers will be the ones referencing it.
